zcash_client_backend/src/keys.rs

```diff
@@ -137,38 +140,135 @@ impl UnifiedSpendingKey {
         if sapling.is_none() {
             None
         } else {
             Some(UnifiedFullViewingKey {
-                account,
                 #[cfg(feature = "transparent-inputs")]
                 transparent,
                 sapling,
+                orchard,
+                // We don't allow constructing new UFVKs with unknown items, but we store
+                // this to allow parsing such UFVKs.
+                unknown: vec![],
             })
         }
     }
 
-    /// Returns the ZIP32 account identifier to which all component
-    /// keys are related.
-    pub fn account(&self) -> AccountId {
-        self.account
+    /// Parses a `UnifiedFullViewingKey` from its [ZIP 316] string encoding.
+    ///
+    /// [ZIP 316]: https://zips.z.cash/zip-0316
+    pub fn decode<P: consensus::Parameters>(params: &P, encoding: &str) -> Result<Self, String> {
+        let (net, ufvk) = unified::Ufvk::decode(encoding).map_err(|e| e.to_string())?;
+        let expected_net = params_to_network(params);
+        if net != expected_net {
+            return Err(format!(
+                "UFVK is for network {:?} but we expected {:?}",
+                net, expected_net,
+            ));
+        }
+
+        let mut orchard = None;
+        let mut sapling = None;
+        #[cfg(feature = "transparent-inputs")]
+        let mut transparent = None;
+
+        // We can use as-parsed order here for efficiency, because we're breaking out the
+        // receivers we support from the unknown receivers.
+        let unknown = ufvk
+            .items_as_parsed()
+            .iter()
+            .filter_map(|receiver| match receiver {
+                unified::Fvk::Orchard(data) => orchard::keys::FullViewingKey::from_bytes(data)
+                    .ok_or("Invalid Orchard FVK in Unified FVK")
+                    .map(|addr| {
+                        orchard = Some(addr);
+                        None
+                    })
+                    .transpose(),
+                unified::Fvk::Sapling(data) => {
+                    sapling_keys::DiversifiableFullViewingKey::from_bytes(data)
+                        .ok_or("Invalid Sapling FVK in Unified FVK")
+                        .map(|pa| {
+                            sapling = Some(pa);
+                            None
+                        })
+                        .transpose()
+                }
+                #[cfg(feature = "transparent-inputs")]
+                unified::Fvk::P2pkh(data) => legacy::AccountPubKey::deserialize(data)
+                    .map_err(|_| "Invalid transparent FVK in Unified FVK")
+                    .map(|tfvk| {
+                        transparent = Some(tfvk);
+                        None
+                    })
+                    .transpose(),
+                #[cfg(not(feature = "transparent-inputs"))]
+                unified::Fvk::P2pkh(data) => {
+                    Some(Ok((unified::Typecode::P2pkh.into(), data.to_vec())))
```

Comment:
Ok it looks as though when `transparent-inputs` is disabled this will treat the P2PKH component as though it were unknown. If that's true then feel free to resolve my questions above.